File: pipgraph-obsidian/src/backend/status.ts
/**
 * Episodic `status` taxonomy — the client mirror of the backend's
 * `app/services/jobs/status.py`. Keep the two in lockstep.
 *
 * Encoding (Variant B): an active job sets `status` to the job's type key; a
 * failure sets `failed:<job_type>`; a settled node has no status. The plugin
 * rarely needs the specific job type — for polling it only asks "in flight,
 * failed, or settled?" — so the predicates below stay job-agnostic.
 */

// Job-runner type keys (also the active `status` value while that job runs).
export const JOB_GENERATE_NAME = "generate_episode_name";
export const JOB_PROCESS_EXISTING = "process_existing_episode";

export const FAILED_PREFIX = "failed:";

/** True if `status` marks a terminal failure (`failed:<job>`). */
export function isFailedStatus(status: string | null | undefined): boolean {
  return typeof status === "string" && status.startsWith(FAILED_PREFIX);
}

/** True if `status` marks an active (non-failed, non-settled) job. */
export function isInFlightStatus(status: string | null | undefined): boolean {
  return typeof status === "string" && status.length > 0 && !isFailedStatus(status);
}

/** True if the node has settled — no job in flight (status absent/empty). */
export function isSettledStatus(status: string | null | undefined): boolean {
  return !status;
}

File: backend/app/services/jobs/status.py
"""Episodic ``status`` taxonomy for the job-runner.

The ``status`` property on an Episodic is the durable, server-side record of
*what async work is in flight for this node* (process-queue plan, P2). It is the
single source of truth that the polling endpoint, the client reconcile, and the
Phase-3 server re-enqueue all read.

Encoding (Variant B — job kind carried in the value, not a second field):

- **Active job**  → ``status`` equals the **job-runner type key** itself
  (``"generate_episode_name"`` / ``"process_existing_episode"``). This makes
  re-enqueue a direct map: ``enqueue(node.status, …)`` with no lookup table.
- **Failed job**  → ``"failed:<job_type>"`` (e.g. ``"failed:process_existing_episode"``).
  The ``failed:`` prefix is the terminal marker; the suffix preserves which job
  failed so a retry knows what to re-run.
- **Settled**     → property absent (``None``). No job in flight.

Why bare ``"processing"`` was dropped: it collided across the two job kinds
(naming vs heavy extraction), so a status-filtered list or a restart re-enqueue
could not tell which job a stuck node needed.
"""

# Job-runner type keys. These double as the *active* ``status`` value of a node
# whose job is in flight (see module docstring). Keep in sync with the handler
# registrations in ``queue.py``.
JOB_GENERATE_NAME = "generate_episode_name"
JOB_PROCESS_EXISTING = "process_existing_episode"

FAILED_PREFIX = "failed:"


def failed(job_type: str) -> str:
    """Status value for a failed job of the given type (``"failed:<job_type>"``)."""
    return f"{FAILED_PREFIX}{job_type}"


def is_failed(status: str | None) -> bool:
    """True if ``status`` marks a terminal failure."""
    return status is not None and status.startswith(FAILED_PREFIX)


def is_in_flight(status: str | None) -> bool:
    """True if ``status`` marks an active (non-failed, non-settled) job."""
    return status is not None and not status.startswith(FAILED_PREFIX)


def job_type_of(status: str | None) -> str | None:
    """The underlying job-runner type for a status, stripping any ``failed:``.

    ``None`` for a settled (absent) status. Used by the Phase-3 re-enqueue to map
    a stuck node back to the handler that should re-run it.
    """
    if status is None:
        return None
    if status.startswith(FAILED_PREFIX):
        return status[len(FAILED_PREFIX):]
    return status
